src/lib/utils.cc

```diff
@@ -682,4 +724,247 @@ TEST_CASE("get_nuc") {
+                data.fsts[i].SetFinal(end - 3);
+            }
+        } else {
+            data.stops.emplace_back("");
+        }
+    }
+}
+
+/// @private
+// GCOVR_EXCL_START
+TEST_CASE("trim_end_stops") {
+    auto test = [](const std::vector<std::string>& raw_seqs,     // NOLINT
+                   const std::vector<std::string>& exp_seqs,     // NOLINT
+                   const std::vector<std::string>& exp_stops) {  // NOLINT
+        coati::data_t data;
+        data.seqs = raw_seqs;
+        for(size_t i = 0; i < raw_seqs.size(); ++i) {
+            data.names.push_back("seq_name");
+        }
+
+        trim_end_stops(data);
+
+        CHECK_EQ(data.seqs, exp_seqs);
+        CHECK_EQ(data.stops, exp_stops);
+    };
+    test({"AAA", "CCC"}, {"AAA", "CCC"}, {"", ""});              // no stops
+    test({"AAATAA", "AAATTT"}, {"AAA", "AAATTT"}, {"TAA", ""});  // stop on ref
+    test({"AAATTT", "AAATAG"}, {"AAATTT", "AAA"}, {"", "TAG"});  // stop on des
+    test({"AAATGA", "AAATGA"}, {"AAA", "AAA"}, {"TGA", "TGA"});  // stop on des
+    test({"AAATAA", "AAATAG"}, {"AAA", "AAA"}, {"TAA", "TAG"});  // stop on des
+
+    auto test_tri = [](const std::vector<std::string>& raw_seqs,     // NOLINT
+                       const std::vector<std::string>& exp_seqs,     // NOLINT
+                       const std::vector<std::string>& exp_stops) {  // NOLINT
+        coati::data_t data;
+        data.seqs = raw_seqs;
+        for(const auto& seq : raw_seqs) {
+            data.names.push_back("seq_name");
+            VectorFstStdArc fsa;  // FSA input sequence
+            coati::acceptor(seq, fsa);
+            data.fsts.push_back(fsa);
+        }
+
+        trim_end_stops(data);
+
+        CHECK_EQ(data.seqs, exp_seqs);
+        CHECK_EQ(data.stops, exp_stops);
+
+        for(size_t i = 0; i < data.size(); ++i) {
+            VectorFstStdArc expected;
+            acceptor(data.seqs[i], expected);
+            CHECK(fst::Equal(expected, data.fsts[i]));
+        }
+    };
+    test_tri({"AAA", "CCC"}, {"AAA", "CCC"}, {"", ""});  // no stops
+    // stop on reference/ancestor
+    test_tri({"AAATAA", "AAATTT"}, {"AAA", "AAATTT"}, {"TAA", ""});
+    // stop on descendant
+    test_tri({"AAATTT", "AAATAG"}, {"AAATTT", "AAA"}, {"", "TAG"});
+    // stop on both sequences
+    test_tri({"AAATGA", "AAATGA"}, {"AAA", "AAA"}, {"TGA", "TGA"});
+    test_tri({"AAATAA", "AAATAG"}, {"AAA", "AAA"}, {"TAA", "TAG"});
+}
+// GCOVR_EXCL_STOP
+
+/**
+ * @brief Restore end stop codons post pairwise alignment.
+ *
+ * @details Four case scenarios:
+ *  (1) no end stop codons on either sequence: nothing to do.
+ *  (2) only present in one sequence: add codon back and insert 3 gaps to other
+        sequence.
+ *  (3) present in both and same stop codon: add codons back as 3 matches.
+ *  (4) present in both but different stop codon: add codons as a
+ *      3 len insertion + a 3 len deletion.
+ *
```

Comment:
For case (4), I think it would be most parsimonious to insert them back as 3 matches. Stop codons do evolve.

Reply:
Makes sense, I agree.